docs/messaging/messaging-spans.md

```diff
@@ -196,21 +198,74 @@ Examples:
+| `create`       | A message is created. "Create" spans always refer to a single message and are used to provide a unique creation context for messages in batch publishing scenarios. |
+| `receive`      | One or more messages are requested by a consumer. This operation refers to pull-based scenarios, where consumers explicitly call methods of messaging SDKs to receive messages. |
+| `deliver`      | One or more messages are passed to a consumer. This operation refers to push-based scenarios, where consumer register callbacks which get called by messaging SDKs. |
+
+### Span kind
+
+[Span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind)
+SHOULD be set according to the following table, based on the operation a span describes.
+
+| Operation name | Span kind|
+|----------------|-------------|
+| `publish`      | `PRODUCER` if the context of the "Publish" span is used as creation context. |
+| `create`       | `PRODUCER` |
+| `receive`      | `CONSUMER` |
+| `deliver`      | `CONSUMER` |
+
+For cases not covered by the table above, the span kind should be set according
+to the [generic specification about span kinds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#spankind),
+e. g. it should be set to CLIENT for the "Publish" span if its context is not
+used as creation context and if the "Publish" span models a synchronous call to
+the intermediary.
+
+Setting span kinds according to this table ensures that span links between
+consumers and producers always exist between a PRODUCER span on the producer
+side and a CONSUMER span on the consumer side. This allows analysis tools to
+interpret linked traces without the need for additional semantic hints.
+
+### Trace structure
+
+#### Producer spans
+
+"Publish" spans SHOULD be created for operations of providing messages for
+sending or publishing to an intermediary. A single "Publish" span can account
+for a single message, or for multiple messages (in the case of providing
+messages in batches). "Create" spans MAY be created. A single "Create" span
+SHOULD account only for a single message. "Create" spans SHOULD either be
+children or links of the related "Publish" span.
+
+If a "Create" span exists for a message, its context SHOULD be injected into
+the message. If no "Create" span exists and no custom creation context is
+injected into the message, the context of the related "Publish" span SHOULD be
+injected into the message.
+
+#### Consumer spans
+
+"Deliver" spans SHOULD be created for operations of passing messages to the
+application when those operations are not initiated by the application code
+(push-based scenarios).
+
+"Receive" spans SHOULD be created for operations of passing messages to the
```

Comment:
For a system that polls but finds no messages, do you get a Receive Span / Independent Trace?

Sorry if this is a silly question, trying to understand how you would be observing the cost of polling, e.g. how many times you try to receive but find no messages available.  Perhaps that's handled on the metrics side?

Reply:
Good question, the answer is two times "yes".

1. Yes, it's planned to have metrics that cover this: https://github.com/open-telemetry/semantic-conventions/pull/163/files#diff-1c3afdf1b3b840fc9e4c17a2110573d1656b7e27d75b13f4ce97e2daff22517bR13-R14

2. Yes, if you initiate a pull-operation and you don't receive any message, you should still get a span. See this text two paragraphs lower:

> A single "Deliver" or "Receive" span can account for a single message, for a batch of messages, or for no message at all (if it is signalled that no messages were received).